flutter/lib/ui/geometry.dart

```diff
@@ -836,40 +1079,89 @@ class RRect {
     return w < h ? w : h;
   }
 
-  /// The point halfway between the left and right and the top and bottom edges of this rectangle.
+  /// The point halfway between the left and right and the top and bottom edges
+  /// of this rectangle.
   Point get center => new Point(left + width / 2.0, top + height / 2.0);
 
-  /// Whether the given point lies inside the rounded rectangle.
+  double _getMin(double min, double radius1, double radius2, limit) {
+    double sum = radius1 + radius2;
+    if (sum > limit && sum != 0.0) {
+      return math.min(min, limit / sum);
+    }
+    return min;
+  }
+
+  /// Scales all radii to the on each side the radii sum will not pass the size
+  /// of the width/height.
+  ///
+  /// Inspired from:
+  ///   https://github.com/google/skia/blob/master/src/core/SkRRect.cpp#L164
+  void _scaleRadii() {
+    if (_scaled == null) {
+      double scale = 1.0;
+      final List<double> scaled = new List.from(_value);
+
+      scale = _getMin(scale, scaled[11], scaled[5], height);
+      scale = _getMin(scale, scaled[4], scaled[6], width);
+      scale = _getMin(scale, scaled[7], scaled[9], height);
+      scale = _getMin(scale, scaled[8], scaled[10], width);
+
+      if (scale < 1.0) {
+        for (int i = 4; i < _kDataSize; i += 1)
+          scaled[i] *= scale;
+      }
+
+      _scaled = new RRect._fromList(scaled);
```

Comment:
caching this won't work, because you don't invalidate the cache when the values change.